src/assign.c

```diff
@@ -408,8 +409,10 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values, SEXP v
     else colnam = STRING_ELT(newcolnames,coln-length(names));
     if (coln+1 <= oldncol && isNull(thisvalue)) continue;  // delete existing column(s) afterwards, near end of this function
     if (vlen<1 && nrow>0) {
-      if (coln+1 <= oldncol) {
-        error("RHS of assignment to existing column '%s' is zero length but not NULL. If you intend to delete the column use NULL. Otherwise, the RHS must have length > 0; e.g., NA_integer_. If you are trying to change the column type to be an empty list column then, as with all column type changes, provide a full length RHS vector such as vector('list',nrow(DT)); i.e., 'plonk' in the new column.", CHAR(STRING_ELT(names,coln)));
+      if (coln+1 <= oldncol) { 
+        if(numToDo > 0){ //fixes #2829
```

Comment:
It took me a while to grok numToDo, but yes.  Looks good.  Could if be one if rather than two? 
`if (coln+1<=oldncol && numToDo>0) {  // numToDo fixes #2829. See test 1911`

Reply:
It can be changed, but then the `else if` statement below needs modification: `else if(coln+1 > oldncol && TYPEOF(thisvalue)!=VECSXP)`. I think, this makes sense and have implemented it in the latest commit.